KPIV-52: Add pivot APIs handler

dhuynh-Kazoo opened the pull request:
New pivot api for `/pivot` and `/pivots/pivotId` with supporting document CRUD actions

kdamas93 committed:
KCRO-148: standardize handling of different types of UI apps (#374)

The purpose of this API is to replicate the behavior of the apps_store
API with the following differences:

- It will be possible to perform actions on the following resources
through the cb_application endpoint
- application document which includes screenshots, icons, entitlement,
and blocklist among others.

The creation and management of permissions for this will be in each
account, unlike the current scheme that was only allowed in the master
account.

This API will also allow handling different types of applications, for
example, commland_ui and monster_ui, a field has been added which
specifies the type of application that has been created. With this it
is hoped to be able to have controlled and grouped the different
applications that may arise within their category.

It has created its own module in core to execute actions from the
console, the module is called kazoo_ui_apps_maintenance, and the
related pr is 2600hz/kazoo-core#524, it has been developed based on
kazoo_documents_maintenance, keeping the functionality that exists for
monster-ui applications that follow the current architecture.

icehess committed:
KCRO-181: ignore empty mms body while there is a atleast a file (#463)

This will remove minimum from MMS schema body field and improves the
cb_mms to make sure we ignore empty body ONLY when there is at least a
file in the request.

If no file and empty body then the request fails as before with
validation error.

Commit:
KZOO-301: Add a period of user lockout for failed login attempts (#467)

Add simple account lock-out mechanism using Token Buckets if any of
the account's users fail to login with correct creds in a configurable
number of attempts.

jamesaimonetti committed:
tinkering with rebar3 as alternative experimental build system (#499)

This introduce rebar3 as an alternative build system, only for use
locally in your dev env.

Current Makefile system is robust and gets the job done. But it
increasingly getting difficult to run this in different env and OS,
and it is ridiculously SLOW. Average time is something around
7mins. If you run in VM it may be more close to 10min. If you add
macOS in equation, it is infinity (something around 25-3m min).

With Rebar3 the build time is something under or around 2min, for
everything, including deps.

Currently this is just a simple configuration to test out
rebar. `xref`, `dialyzer`, test and other targets are not tested, and
are not properly configured.

This tries to not depend on calling Makefile targets as much as
possible. This reduce the problems on different platforms especially
windows and macos. Some scripts was added to replace some tasks needed
to setup files for compilation.

jamesaimonetti committed:
KZOO-397: update schema for expected format of classifiers (#494)

Include required classifier properties

Commit:
PS-137: Doorman callflow schema updates (#496)

Commit:
PS-137: Doorman update based on recent changes made for customer (#535)

Commit:
KPUS-6: add push-type to AMQP payload (#495)

icehess committed:
bump orb to use custom couchdb image (#498)

kusalfernando committed:
KCRO-183: multiple functions to handle all possible scenarios for direction (#481)

* break direction/2 into multiple functions to handle all possible scenarios

* use proper function names and add comments to clarify

---------

Co-authored-by: Hesaam Farhang <hesaam@2600hz.com>

kdamas93 committed:
KCRO-188: Update contact lists (#500)

The purpose of this PR is to improve/adapt the current functionality
of cb_list to be able to save contact lists depending on the needs,
with these changes it will be possible to associate owner_id to the
lists so that they can later be filtered by user.

With this owner_id permission, validation logic will be added, about
who can access the lists.

Multiple entries can be associated to a list type document, each entry
will be a contact list, to achieve this it is necessary to update the
current schema (list_entry), the documentation and its respective
accessors.

It is also necessary to perform a refactor on the current code because
there are functions/calls that do not follow our current programming
patterns.

icehess committed:
update orb to consider os-release in cache keys (#502)

With upcoming changes, we need another variable to invalidate caches.

This adds /etc/os-release, uname -pmo and others together and uses them
as cache key in CI.

kdamas93 committed:
kcro-148:  allow admin to list available apps to their account (#506)

* KCRO-148: allow admin to list available apps to their account

This enables admin to see what is available so they can create
entitlement for it which enables the app in their account.

* update doc

---------

Co-authored-by: Hesaam Farhang <hesaam@2600hz.com>

Navoda626 committed:
KCRO-190: Additional APIs to consume pvt_chat.getstream in user doc (#505)

* Adding enable getstream feature and removing getstream feature APIs

* Adding APIs for geting user getstream enabled status and account getstream enabled users

* changes after ci checks

* Added documentation and removing the API for fetching getstream status for an user since its covered by user fetch by id API

* adding API to get the user getstream status

* Adding getstream accessor functions and changing get user getstream to pass the getstream obj

* enable adding json body for the getstream object for the PUT request

* Updating the DELETE method to remove getstream obj and pass the removed obj

* Adding patch method for getstream

* After circle-ci tests

* removing unwanted config file of teammate application

* after circle-ci checks

* refactor folds into manual recursion loops

* after code review

* adding meaningful error message

* resolving code-checks

* handling response codes from getstream API

* fetch getstream enabled users using crossbar_listing and erlang filtering

* removing unwanted log

* code cleanups

* code cleanups

---------

Co-authored-by: Navoda626 <navoda@aio.kazoo150.com>
Co-authored-by: James Aimonetti <james@2600hz.com>

jamesaimonetti committed:
AXS-101: schema updates (#514)

jamesaimonetti committed:
Schema updates from other public apps' cb modules

Navoda626 committed:
KCRO-189: schema updates for whitelabel (#508)

Adding new commland object and parameters in whitelabel schema

kdamas93 committed:
KCRO-194: Fix the double URL encoding (#518)

kdamas93 committed:
KZOO-410: UUID treatment for command selector (#516)

Co-authored-by: Hesaam Farhang <icehess@users.noreply.github.com>

kdamas93 committed:
KCRO-148: Updating docs with the correct path (#520)

Fixing patch of an app to update the document correctly

manushiperera committed:
KCRO-195 : As a user I want to update data from the Monster UI Marketplace connector (#522)

* Previously it was unable to set is_aio_cluster configuration through PATCH apps_store/marketplace API. Now this configuration can be set by PATCH apps_store/marketplace: {action:enable, settings:[is_aio_cluster,…]}. settings is an optional parameter.
A new market action is defined to handle the settings payload. Cluster name is added to response payload and cluster name can be updated via marketplace connector.

* explicitly handle updating local vs remote settings

* update docs, add more

* Fixed functionality issues, CI issues and typos
Remove api_url market action as it is depreciated

---------

Co-authored-by: Hesaam Farhang <hesaam@2600hz.com>
Co-authored-by: Hesaam Farhang <icehess@users.noreply.github.com>

kdamas93 committed:
KCRO-197: whitelabeled/blacklisted apps breaks the app selector list (#523)

Adjusting instructions for managing applications that come from the service plan
When a service plan has a whitelabeled/blacklisted application generates a break since the structure returned is not what was expected, this exception is adapted with the added changes
error reported :
https://2600hz-commercial.atlassian.net/browse/CAS-110

dhuynh-Kazoo committed:
KZOO-301: Set infinity for cb_conference_auth for account level (#527)

harenson committed:
DOC: Add guide to setup a local S3-compatible service MinIO (#530)

kdamas93 committed:
KCRO-188: Adding feature parity with the new fields of the contacts object (#532)

-It is necessary to be able to provide compatibility to the new fields
that will be used within the contacts object

-Currently, there is no way to fetch all the needed data from all the
lists at the same time I had to fetch the data from every list one by
one

lazedo committed:
KCAL-58 update callflow.park schema

icehess committed:
AXS-133: add a cb module to handle ui wrapper app routes (#540)

Currently UI asset wrapper app uses Crossbar Custom Route to add their
routes to the Crossbar. But this needs a restart of crossbar to take
effect.

It is not useful to sysadmin as they start any UI app wrapper from
Marketplace and they expect its assets be available from a currently
running crossbar without any interruption to API availability.

This adds a new module to crossbar to handle `/ui/[...]` routes and
calls the apps wrapper module if it is exists.

The first part of path immediately after `ui` is the name of UI
wrapper app. The rest is the path to assets itself.

`crossbar_ui_wrapper_handler` will register itself to `/ui/[...]` in
cowboy router, and upon its calling, checks the app has a specific
module to handle this route is loaded then upgrades to that module for
handling the request.

For example:

`/ui/monsterui_callflows/[...]` will init
`crossbar_ui_wrapper_handler` then it tries to upgrade the handler to
`monsterui_callflows_cb_router` module if it exists. if not 404 will
be returned.

icehess committed:
AXS-134: filter out ui apps started from marketplace if its app is not running (#543)

When Apps Store API or the new Applications API is returning the list
of allowed or available apps, they should now check if the app is from
Marketplace, and if yes they have to check whether its app is started
or not.

This will make sure these APIs only returning the apps that are
accessible and running in the system.

manushiperera committed:
KTEL-21: Add DKIM support to whitelabel settings (#545)

A private key file and DKIM selector information are added to the
whitelabel API. A new API is defined to upload the private key
file. If a file is already available, it is replaced. Otherwise the
file is saved and put the file ID in whitelabel doc as a pvt field.

lazedo committed:
add presence aliases schema updates

Commit:
KCRO-202: Refactor the cb_lists endpoint for one to one support for entry and identity (#560)

Previously, cb_lists endpoints are not utilized to manage contact
entries properly. Now, cb_lists endpoints are refactored for both
company contacts which are not associated with user ids and personal
contacts which are associated with user ids. Further permissions and
ownerships are introduced to access the list of contacts by users.

jamesaimonetti committed:
Remove the stacktrace macro

jamesaimonetti committed:
Update whitelabel doc

swysor committed:
Test Commit

danielfinke committed:
Delete unused `kapi_money` (#550)

* Delete unused `kapi_money`

* Bump copyright year in `crossbar_app`

danielfinke committed:
KWEB-12: use `kapi_webhooks` for webhook re-enablement req (#553)

danielfinke committed:
KCRO-205: fix `cb_user_auth` always logging "found more than one user" (#554)

danielfinke committed:
KBLA-15: add BH object binding configs for allowed doc types (#555)

* KBLA-15: add BH object binding configs for allowed doc types

* KBLA-15: address CI `docs` failures

icehess committed:
KZOO-447: maintain reseller id for descendants when moving a reseller account (#557)

Update how we set the new reseller id for descendant accounts when a
reseller account is being move under new account tree.

Previously the reseller id for newly moved and all of its descendants
would be set to the destination account's reseller id. This old way
would attentionally will end up descendants to have super duper admin
as their new reseller.

This changes that when moving an account:

1. For the account that being moved, we look at the destination
account, if the destination is a reseller account, then we use this a
new reseller id, if not then we lookup the reseller id of destination

2. For descendants of account, we first check if the account is
reseller itself, if yes we set the descendants reseller id to be the
account itself, if not then new reseller id of the account will be
used.

jamesaimonetti committed:
KZOO-450: add emergency address fields for new provider (#567)

manushiperera committed:
KCRO-206: DKIM keys per from email address to sign notification emails (#580)

Previously, DKIM information was common for the account and unable to
configure based on the available email addresses of the account.

Now, private key file and DKIM selector information can be configured
for each email address through new whitelabel API.

CRUD operations are defined to configure, verify and update email
addresses along with DKIM information.

jamesaimonetti committed:
ignore test (doesn't seem particularly helpful) (#575)

Part of OTP 26 work, this test module has two tests failing for "reasons". I don't think they're particularly helpful tests anyway.

* ignore test (doesn't seem particularly helpful)
* edoc and bump

icehess committed:
Schema properties must specify their type (#855)

jamesaimonetti committed:
OTP-26: Update parameters for mimemail (#577)

Commit:
update shipyard config to support rockylinux

Commit:
enable circle-ci support for rocky linux

k-anderson committed:
KCRO-230: added ability to create draft ledgers via the API (#631)

jamesaimonetti committed:
Update cnam system_config schema

jamesaimonetti committed:
Splchk of 'perform' variations

jamesaimonetti committed:
Splchk of 'transferring' variants

jamesaimonetti committed:
KZOO-498: include user_id in mfa 401 (#401)

When MFA'd auth request is 401'd to require the MFA (meaning the
password hash is for a valid user in the account but no MFA token is
present), include the user_id of the requestor if available.

danielfinke committed:
Use OTP 25 `lists:uniq/1` (#584)

icehess committed:
KPIV-17: add leave_endconf flag and endconf_grace_time (#574)

* add leave_endconf flag and endconf_grace_time

Adding per member/modrator options to end conference when last member with
that flags exists the conference.

Also adds the endconf_grace_time.

jamesaimonetti committed:
KZOO-503: use "format" for SIP hostnames (#597)

Previously we just used the kz_json_schema_extensions is_valid_host/1
to check the value of a gateway's 'server'. If the domain provided did
not configure an A record in DNS, the schema validation would fail.

Since SIP clients can use SRV to get alternative hosts/IPs, failing
validation wasn't correct.

With changes to core, validation of hostnames take into account SRV
records.

Hostname validation is also documented on the resources.md

jamesaimonetti committed:
Use python3 to format JSON

jamesaimonetti committed:
Include uuid value under call_id key (#598)

jamesaimonetti committed:
KZOO-498: fix missing change for POST (#599)

PATCH was updated with new return from
maybe_add_multi_factor_metadata/1 but the POST handler was not.

jamesaimonetti committed:
KCRO-216: check for user's password expiration (#604)

- include password expiration timestamp in user metadata
- set require_password_update to true if password is expired

jamesaimonetti committed:
KCRO-216: don't set require_password_update by default

jamesaimonetti committed:
KCRO-216: Fetch user using kzd_users module

jamesaimonetti committed:
Just set pvts on config doc and be done (#605)

When diffing a doc that has a PATCH value equal to the config doc's
value (such as PATCHing "token_cost":1 when the config doc already has
"token_cost":1), the unchanged key ("token_cost") was stripped (since
no diff in value) and not saved.

icehess committed:
rebar3: add hooks to generate timezone and deps license list (#606)

jamesaimonetti committed:
Correct spelling of mismatch (#608)

wdhammika committed:
KCRO-218: Failure to Save Message Metadata Due to Incorrect Context Status Check in cb_vmboxes Module (#609)

Previously, it would fail when we tried to save voicemail media messages.
Now: Handle error causes that cause problems in saving records. (Remove extra validations)

* handle vmbox media upload error

* bump copyright year

* check response status in put/3 funtion

* code check issue fixed

---------

Co-authored-by: wimal <wimal@kazoo.vm>

jamesaimonetti committed:
KCRO-227: Parse X-Forwarded-For header for client IP (#615)

dhuynh-Kazoo committed:
DDOC-67: Fix formatting of enum types in JSON schema tables (#572)

enoskov-at-ooma committed:
KCRO-225 Need to support Call Recording on demand (#614)

Add support to channels API to start and stop call recording.
This is used by Ooma mobile and desktop clients.

Co-authored-by: Evgeny Noskov <evgeny.noskov@ooma.com>

jamesaimonetti committed:
KZOO-543: add schema update

eezhov-ooma committed:
KZOO-537: Shared VM Box (#625)

Update schema for vmbox doc

k-anderson committed:
KBRA-3: schema updates for new braintree system_config parameters (#629)

dhuynh-Kazoo committed:
KPIV-51: Add auto-record and recording-data to conference.json (#613)

Add auto-record and recording-data to conference.json

* Adding auto-record and recording-data to conference.json
* update schemas
* update schemas

---------

Co-authored-by: Hesaam Farhang <hesaam@2600hz.com>

icehess committed:
KPIV-47: adding pivot stream system_config json schema (#633)

Related to https://github.com/2600hz/kazoo-pivot/pull/48

manushiperera committed:
KZOO-532: Track manual presence in separate docs vs one master doc (#636)

Previously, manual presence records were maintained in a single
document per master account.

Now, the system has been enhanced to track manual presence in separate
documents for each presence ID, stored under the modb database.

At the end of each month, these manual presence documents are rolled
over to the following month's modb for continuity and efficient
management.

jamesaimonetti committed:
KZOO-166: Use kzd module for paths (#639)

jamesaimonetti committed:
KZOO-166: fix path modified key

Commit:
KZOO-531 Call recording status through API

icehess committed:
KPIV-53: support Pivot serverless processing of a flow (#650)

Callflow Pivot module can now accepts a serverless field
and provides the Pivot flow to be executed.

This removes the dependence of a web server and make using
Pivot serverless possible.

The new `serverless` field is an object with `flow_doc` and `flow_type`
fields.

- `flow_doc`: can be string with a valid PivotML (a twiml extension) or
  a valid Callflow JSON
- `flow_type`: indicates the type of specified flow_doc. can be aither
  `application/json` for callflow or `application/xml` or `text/xml` for
  PivotML.

Using `serverless` or `voice_url` is mutual exclusive, meaning they
can't be used together. In serverless special take must be taken to
use a full URL for any action url throughout the PivotML and relative
URL will not work in serverless.

Flow doc will be serialized to string if it is JSON and then will be send
to Pivot app to be processed alongside the flow_type.

Pivot will start processing the flow-doc if any, otherwise expects
voice_url. The execution will proceed as usual (other than expected
full URL for all action URLs as mentioned above).

wdhammika committed:
update descendants count after moving accouts (#646)

Co-authored-by: wimal <wimal@kazoo.vm>

icehess committed:
DUPT-566: fix load_user won't store db_doc (#658)

The way the user doc is read is changed, without storing the original doc
stored in db_doc causing any code that depends on that (welcome
notification for example) to fail.

Also changing the code when webphone enabled notification is sent, to
send it only when a webphone device is set or its current value is
changed.

jamesaimonetti committed:
KCRO-216: check user password expiration when authing (#660)

lazedo committed:
fail_on_single_reject is boolean or list of causes

dhuynh-Kazoo committed:
KCRO-236: sync crossbar with provisioner v6 (#654)

- Add POST /accounts/{account_id} from kazoo -> provisioner when a new account is created
- Add provision template id to be compatible with provisioned v6

jamesaimonetti committed:
KZOO-584: update oauth settings for new provider (#665)

client secret is not required for some providers
include code_verifier field in callback schema

pavel-farkas committed:
KZOO-586: Validation for media_name field of metaflows.record_call schema (#667)

Co-authored-by: Pavel Farkas <pfarkas@2600hz.com>

dhuynh-Kazoo committed:
KPIV-71: Add auto-record and min-required-recording-participants to conferences.profile.json (#663)

- the auto-record and min-required-auto-recording-participants are set from conference doc. Now they are removed from the conference doc in this PR.  So these fields should be configured in the account conference config profile (configs_conferences doc).
- update  auto-record and min-required-auto-recording-participants in conferences.profile.json

romanat-ooma committed:
KCRO-240: fix log message (#676)

jamesaimonetti committed:
KZOO-584: use access_token as auth token (#678)

* KZOO-584: use access_token as auth token

* fix return when auth token is not a JWT

lazedo committed:
KCAL-71 add property to callflow.user

romanat-ooma committed:
KZOO-589: Update docs with new value and remove deprecated values (#673)

* update docs with new value and remove deprecated values

* set default to fetch

filipevarjao committed:
adding a note to refresh/renew auth token doc (#682)

romanat-ooma committed:
KZOO-576: Disable auto indexing of couchdb views (#688)

This adds schema for design_doc system_config for kazoo_data.

---------

Co-authored-by: Hesaam Farhang <hesaam@2600hz.com>

dhuynh-Kazoo commented:
master:
https://github.com/2600hz/kazoo-crossbar/pull/687
https://github.com/2600hz/kazoo-callflow/pull/295
https://github.com/2600hz/kazoo-core/pull/1137
https://github.com/2600hz/kazoo-properly/pull/239
https://github.com/2600hz/kazoo-ast/pull/55

5.4:

lazedo committed:
KCRO-244 implement progress routines for modules to use on long running operations

lazedo committed:
KCRO-245 use progress for account deletion

* deleting an account can be a long-running procedure and we need to give feedback to client so it doesn't time out

icehess commented on priv/couchdb/schemas/callflows.pivot.json:
```diff
@@ -5,111 +5,16 @@
-            "kazoo-validation": true,
-            "minLength": 7,
-            "pattern": "^https?://",
+        "id": {
```
Can you move `id` to `oneOf` first object and remove properties here? it is more cleaner this way.

icehess commented on doc/pivot.md:
```diff
@@ -23,6 +23,240 @@ The Pivot endpoint is not loaded on start in a default Kazoo installation.
 See the [Pivot callflow schema](../../callflow/doc/pivot.md) for details.
 
+## Fetch all pivot flows for an account
+
```
Please for new code, write proper documentation. Explain this endpoint, what are its uses, and compare the old format and the new format etc...

Also please write properly test for CRUD on pivot. And write simple test for using `id` in pivot callflow.

icehess committed:
KHEL-4: leak signup params as metadata (#693)

Standardizing the private sign up fields, so each method or service that
is sued for signing up the account can set some generic parameters and
can be leaked as metadata from API when fetching account.

One of the parameter, `email_is_verified` needs further updates by other
PRs to add the email verification steps, reminder, sending emails and
etc.... The idea is that that the service or the method used for signup
will have to provide those mechanism and set this flag true when email
used for signup is verified. UI can use this generic flag to show a
notice to user that the email is needed to be verified.

Also adding `pvt_is_trial` so and leak it via API, this is still wip,
but currently we needed a way in CPAAS to determine an account is still
in trial mode when using a feature we don't trail account have access.
Other PRs will add the logic to update and remove this trial flag.

This trial flag is not time-based, at least for intended usage in CPAAS.

lazedo committed:
KCRO-245 add reset_idle_timeout_on_send

* when sending intermediate answers with http status 102 we need to reset idle_timeout timer

lazedo committed:
KCRO-245 allow removing context stored values

lazedo committed:
KCRO-245 add application for logging

lazedo committed:
KCRO-245 ensure intermediate msg is sent

* if one of the steps takes too long we syill need to send intermediate 102 msg to client

dhuynh-Kazoo committed:
KWEB-16: Add custom_http_headers for webhook to webhooks schema (#675)

Add custom_http_headers for webhook request

dhuynh-Kazoo committed:
Add pivot API CRUD and new schema

artable committed:
KWEB-14: (Reliably) send account creation and deletion webhooks (#680)

Crossbar sends `account_deleted` notifications, and also passes `Should-Notify` in `account_created` notifications.

dhuynh-Kazoo committed:
Merge branch 'master' into KPIV-52
